service/cluster/cluster.go

```diff
@@ -0,0 +1,356 @@
+		if ctx.Err() != nil {
+			return false
+		}
+
+		tracer := s.tracer.Tracer("")
+		spanCtx, span := tracer.Start(ctx, "NotifyClusterChange", trace.WithSpanKind(trace.SpanKindInternal))
+		defer span.End()
+
+		names := make([]string, len(peers))
+		for i, p := range peers {
+			names[i] = p.Name
+		}
+		level.Info(s.log).Log("msg", "peers changed", "new_peers", strings.Join(names, ","))
+
+		// Notify dependant components about the clustering change.
+		for _, consumer := range host.GetServiceConsumers(ServiceName) {
+			if ctx.Err() != nil {
+				break
+			}
+
+			clusterComponent, ok := consumer.Value.(Component)
+			if !ok {
+				continue
+			}
+
+			_, span := tracer.Start(spanCtx, "NotifyClusterChange", trace.WithSpanKind(trace.SpanKindInternal))
+			span.SetAttributes(attribute.String("consumer_id", consumer.ID))
+
+			clusterComponent.NotifyClusterChange()
+
+			span.End()
+		}
+
+		return true
+	}))
+
+	peers, err := s.getPeers()
+	if err != nil {
+		return fmt.Errorf("failed to get peers to join: %w", err)
+	}
+
+	if err := s.node.Start(peers); err != nil {
+		level.Warn(s.log).Log("msg", "failed to connect to peers; bootstrapping a new cluster", "err", err)
+
+		err := s.node.Start(nil)
+		if err != nil {
+			return fmt.Errorf("failed to bootstrap a fresh cluster with no peers: %w", err)
+		}
+	}
+
+	if s.opts.RejoinInterval > 0 {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			t := time.NewTicker(s.opts.RejoinInterval)
+			defer t.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+
+				case <-t.C:
+					peers, err := s.getPeers()
+					if err != nil {
+						level.Warn(s.log).Log("msg", "failed to refresh list of peers", "err", err)
+						continue
+					}
+
+					level.Info(s.log).Log("msg", "rejoining peers", "peers", strings.Join(peers, ","))
+					if err := s.node.Start(peers); err != nil {
+						level.Error(s.log).Log("msg", "failed to rejoin list of peers", "err", err)
+						continue
+					}
+				}
+			}
+		}()
+	}
+
+	<-ctx.Done()
+	return nil
+}
+
+func (s *Service) getPeers() ([]string, error) {
+	if !s.opts.EnableClustering || s.opts.DiscoverPeers == nil {
+		return nil, nil
+	}
+	return s.opts.DiscoverPeers()
+}
+
+func (s *Service) stop() {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	// The node is going away. We move to the Terminating state to signal
+	// that we should not be owners for write hashing operations anymore.
+	//
+	// TODO(rfratto): should we enter terminating state earlier to allow for
+	// some kind of hand-off between components?
+	if err := s.node.ChangeState(ctx, peer.StateTerminating); err != nil {
+		level.Error(s.log).Log("msg", "failed to change state to Terminating", "err", err)
+	}
+
+	if err := s.node.Stop(); err != nil {
+		level.Error(s.log).Log("msg", "failed to gracefully stop node", "err", err)
+	}
+}
+
+// Update implements [service.Service]. It is a no-op since the cluster service
```

Comment:
I don't think an error counts as a no-op :) I agree we should have an error though.